### ISSUE TITLE
ValueError('list.remove(x): x not in list',) when creating a security group

### ISSUE THREAD

**Issue by automate-solutions** [inmanta/openstack#5]
When adding a security group to a VM, it fails the security group creation step (even if it does create it correctly on OpenStack) and then skips the rest. 
Here is the error available:
```
An error occurred during deployment of openstack::SecurityGroup[iaas_openstack2,name=mgmt_sg],v=1499093829 (exception: ValueError('list.remove(x): x not in list',))
```
The security group is added in the host implementation (in mymodule):
```
implementation myhost for MyHost:
    #wire up all config for agent injection
    env_name = std::environment_name()
    env_id = std::environment()
    env_server = std::environment_server()
    port = std::server_port()
    #wire up all config for vm creation
    self.provider = cluster.provider
    self.project = cluster.project
    self.image = cluster.image_id
    self.subnet = cluster.subnet
    self.user_data = std::template("mymodule/user_data.tmpl")
    self.key_pair = cluster.key
    self.os = cluster.os
    self.flavor = cluster.flavor
    sg_mgmt = openstack::SecurityGroup(provider=self.provider, project=self.project, name="mgmt_sg", description="Test Mgmt SG")
    openstack::IPrule(group=sg_mgmt, direction="egress", ip_protocol="all", remote_prefix="0.0.0.0/0")
    openstack::IPrule(group=sg_mgmt, direction="ingress", ip_protocol="icmp", remote_prefix="0.0.0.0/0")
    openstack::IPrule(group=sg_mgmt, direction="ingress", ip_protocol="tcp", port=22, remote_prefix="0.0.0.0/0")
    openstack::IPrule(group=sg_mgmt, direction="ingress", ip_protocol="all", remote_prefix="0.0.0.0/0")
    self.vm.security_groups=[sg_mgmt]
end
```
In `main.cf` the hosts are created:
```
host1 = mymodule::MyHost(name="testhost", cluster=cluster)
host2 = mymodule::MyHost(name="host2", cluster=cluster)
```

**Comment by bartv** [comment]
A question and a remark:
 * Did you create the VMs first and added the security group afterwards?
 * You define the security group twice in your model. Because it is the same implementation and a security group has an index on project and name, the compiler will define only one resource. However, we think as a best practice it is cleaner to define that security group somewhere else, for example in the implementation of the cluster entity that I see in your code snippet.

**Comment by automate-solutions** [comment]
• The VMs were created beforehand. I built the model incrementally. Nevertheless, I recreated the environment from the repository and I encountered the same issue.
• I'm making another implementation in a cleaner way. Thanks for the advice

**Comment by bartv** [comment]
I'am unable to reproduce this error. Do you also get this error with a single virtual machine?

**Comment by automate-solutions** [comment]
I have the same behavior. If I deploy only one VM, I don't get the error.
It only appears when I try to apply this SG on two VMs.
Here is the output of the agent log:
```
[root@inmanta-dev tmp]# grep -A 10 -B 10 'list.remove(x)' /var/log/inmanta/agent-e90d5ab3-9b8e-426f-86b0-66cad2beb406.err 
2017-07-06 08:16:56,718 DEBUG   RESP: [200] Content-Type: application/json Content-Length: 23 X-Openstack-Request-Id: req-175f6ece-d75e-43ac-817f-c441322cba18 Date: Thu, 06 Jul 2017 06:16:55 GMT Connection: keep-alive 
RESP BODY: {"security_groups": []}

2017-07-06 08:16:56,718 DEBUG   GET call to network for http://172.16.75.161:9696/v2.0/security-groups?name=mgmt_sg used request id req-175f6ece-d75e-43ac-817f-c441322cba18
2017-07-06 08:16:56,719 DEBUG   REQ: curl -g -i -X POST http://172.16.75.161:9696/v2.0/security-groups -H "Accept: application/json" -H "Content-Type: application/json" -H "X-Auth-Token: {SHA1}fef0cfb96081f031b06e1588bc6198419aac09f1" -H "User-Agent: python-neutronclient" -d '{"security_group": {"description": "Test Mgmt SG", "name": "mgmt_sg"}}'
2017-07-06 08:16:56,847 DEBUG   http://172.16.75.161:9696 "POST /v2.0/security-groups HTTP/1.1" 201 1315
2017-07-06 08:16:56,848 DEBUG   RESP: [201] Content-Type: application/json Content-Length: 1315 X-Openstack-Request-Id: req-133e989e-ab19-4ce9-9b9d-65a4a1c3919c Date: Thu, 06 Jul 2017 06:16:55 GMT Connection: keep-alive 
RESP BODY: {"security_group": {"description": "Test Mgmt SG", "tenant_id": "5878aebf371d4df3aa5bcb0cc6456720", "created_at": "2017-07-06T06:16:55Z", "updated_at": "2017-07-06T06:16:55Z", "security_group_rules": [{"direction": "egress", "protocol": null, "description": null, "port_range_max": null, "updated_at": "2017-07-06T06:16:55Z", "revision_number": 1, "id": "bb08738c-6146-4eb0-afe7-aeca4527cb51", "remote_group_id": null, "remote_ip_prefix": null, "created_at": "2017-07-06T06:16:55Z", "security_group_id": "67e244eb-438b-4069-bb44-bc1f6beaac88", "tenant_id": "5878aebf371d4df3aa5bcb0cc6456720", "port_range_min": null, "ethertype": "IPv4", "project_id": "5878aebf371d4df3aa5bcb0cc6456720"}, {"direction": "egress", "protocol": null, "description": null, "port_range_max": null, "updated_at": "2017-07-06T06:16:55Z", "revision_number": 1, "id": "eec9e016-35c4-4f94-acd8-57328a2f88e1", "remote_group_id": null, "remote_ip_prefix": null, "created_at": "2017-07-06T06:16:55Z", "security_group_id": "67e244eb-438b-4069-bb44-bc1f6beaac88", "tenant_id": "5878aebf371d4df3aa5bcb0cc6456720", "port_range_min": null, "ethertype": "IPv6", "project_id": "5878aebf371d4df3aa5bcb0cc6456720"}], "revision_number": 1, "project_id": "5878aebf371d4df3aa5bcb0cc6456720", "id": "67e244eb-438b-4069-bb44-bc1f6beaac88", "name": "mgmt_sg"}}

2017-07-06 08:16:56,848 DEBUG   POST call to network for http://172.16.75.161:9696/v2.0/security-groups used request id req-133e989e-ab19-4ce9-9b9d-65a4a1c3919c
2017-07-06 08:16:56,848 ERROR   An error occurred during deployment of openstack::SecurityGroup[iaas_openstack,name=mgmt_sg],v=1499321807 (exception: ValueError('list.remove(x): x not in list',))
2017-07-06 08:16:56,849 INFO    end run openstack::SecurityGroup[iaas_openstack,name=mgmt_sg],v=1499321807
```

**Comment by bartv** [comment]
I am unable to reproduce this. Could you provide me with the project and modules? (Remove any openstack credentials in it).

**Comment by automate-solutions** [comment]
I have the following structure:

- libs:
    - mymodule:
        - model:
            - _init.cf
        - templates:
            - user_data.tmpl
        - module.yml
- main.cf
- project.yml

_init.cf:
```
import openstack
import ssh


entity MyCluster:
    """
        A cluster object that represents all shared config and infrastructure,
        including connecting to OpenStack.
    """
    string network_name
    string subnet_name
    string network_address
    string image_id 
    string flavor
end

#input: the ssh key for all VMs
MyCluster.key [1] -- ssh::Key

#input: the OS for all VMs
MyCluster.os [1] -- std::OS

#internal: objects needed to construct hosts
MyCluster.provider [1] -- openstack::Provider
MyCluster.project [1] -- openstack::Project
MyCluster.net [1] -- openstack::Network
MyCluster.subnet [1] -- openstack::Subnet

implementation connection for MyCluster:
    # Define the OpenStack provider to use
    self.provider = openstack::Provider(name="iaas_openstack",
                                        connection_url=std::get_env("OS_AUTH_URL"),
                                        username=std::get_env("OS_USERNAME"),
                                        password=std::get_env("OS_PASSWORD"),
                                        tenant=std::get_env("OS_PROJECT_NAME"))

    # Define the project/tenant to boot the VM in, but do not let inmanta manage it
    self.project = openstack::Project(provider=self.provider, name=self.provider.tenant,
                                      description="", enabled=true, managed=false)

    # Define the network objects to connect the virtual machine to but again, do not manage them
    self.net = openstack::Network(provider=self.provider, project=self.project,
                                  name=self.network_name, managed=false)
    self.subnet = openstack::Subnet(provider=self.provider, project=self.project,
                                    network=self.net, dhcp=true, name=self.subnet_name,
                                    network_address=self.network_address, managed=false)
end

implement MyCluster using connection

#define our own host type
entity MyHost extends openstack::Host:
end

#input: the cluster object
MyCluster.hosts [0:] -- MyHost.cluster [1]

implementation myhost for MyHost:
    #wire up all config for agent injection
    env_name = std::environment_name()
    env_id = std::environment()
    env_server = std::environment_server()
    test = std::server_port()
    port = std::server_port()
    another = std::server_port()
    #wire up all config for vm creation
    self.provider = cluster.provider
    self.project = cluster.project
    self.image = cluster.image_id
    self.subnet = cluster.subnet
    self.user_data = std::template("mymodule/user_data.tmpl")
    self.key_pair = cluster.key
    self.os = cluster.os
    self.flavor = cluster.flavor
    sg_mgmt = openstack::SecurityGroup(provider=self.provider, project=self.project, name="mgmt_sg", description="Test Mgmt SG")
    openstack::IPrule(group=sg_mgmt, direction="egress", ip_protocol="all", remote_prefix="0.0.0.0/0")
    openstack::IPrule(group=sg_mgmt, direction="ingress", ip_protocol="icmp", remote_prefix="0.0.0.0/0")
    openstack::IPrule(group=sg_mgmt, direction="ingress", ip_protocol="tcp", port=22, remote_prefix="0.0.0.0/0")
    openstack::IPrule(group=sg_mgmt, direction="ingress", ip_protocol="all", remote_prefix="0.0.0.0/0")
    self.vm.security_groups=[sg_mgmt]
end

# use our implemenation
# and also the catchall std::hostDefaults
# and the openstackVM implementation that sets the ip and create the eth0 port
implement MyHost using myhost, std::hostDefaults, openstack::openstackVM, openstack::eth0Port
```
user_data.tmpl:
```
#!/bin/bash

hostname {{ name }}
setenforce 0

cat > /etc/yum.repos.d/inmanta.repo <<EOF
[bartvanbrabant-inmanta]
name=Copr repo for inmanta owned by bartvanbrabant
baseurl=https://copr-be.cloud.fedoraproject.org/results/bartvanbrabant/inmanta/fedora-\$releasever-\$basearch/
type=rpm-md
skip_if_unavailable=True
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/bartvanbrabant/inmanta/pubkey.gpg
repo_gpgcheck=0
enabled=1
enabled_metadata=1
EOF

dnf install -y python3-inmanta-agent

cat > /etc/inmanta/agent.cfg <<EOF
[config]
heartbeat-interval = 60
fact-expire = 60
state-dir=/var/lib/inmanta
environment={{ env_id }}
agent-names=\$node-name
[agent_rest_transport]
test={{test}}
host={{env_server}}
EOF

systemctl start inmanta-agent
systemctl enable inmanta-agent
```
module.yml:
```
author: Inmanta <code@inmanta.com>
license: Apache 2.0
name: mymodule
version: 0.1.0
```
main.cf:
```
import ssh
import redhat
import ubuntu

## Edit this parameters
image_id = "52a7e16f-8d48-4369-af78-2da1d23d2011"
network_name = "private"
subnet_name = "private_subnet"
network_address = "10.0.0.0/24"

flavor_name = "m1.little"
ssh_public_key="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC7KC7uZ/31S2AfbFZ+3EL5xL29/AF6N5C5oFLDg0qxJwnbLMY9MmU/QbSYG13YkSIyujEGrOqgJH9Q3S6biQ3ibU2RfmXCm0lnWweayz9eKaBbHMF1CqsYnEe6Mt7r2NeOh2uxeJyko+tIeEvOvwC/JZfVV3tBPej0se20SShG/EPHQ2KjWRMMbDBZ57OZ8MT+xmg5qsWobKJHwoIzFYPwlQ07s0Qe7KGBFxBOp5vogZ9Y162Gas7M2/8uz/YfmaL+JIry5Fl88bBtirIZKro7LOq1pDernoqbHnheJo6PAiFzmyEqYVSG5Vlom6F7Gr0Rgmg4BR525q/ilj5sIC7t root@inmanta"

os = redhat::fedora23
## End edit

# register ssh key
ssh_key = ssh::Key(name="mykey", public_key=ssh_public_key)

# Define the OpenStack provider to use
provider = openstack::Provider(name="iaas_openstack", connection_url=std::get_env("OS_AUTH_URL"),
                               username=std::get_env("OS_USERNAME"),
                               password=std::get_env("OS_PASSWORD"),
                               tenant=std::get_env("OS_PROJECT_NAME"))

# Define the project/tenant to boot the VM in, but do not let inmanta manage it
project = openstack::Project(provider=provider, name=provider.tenant, description="", enabled=true,
                             managed=false)

# Define the network objects to connect the virtual machine to but again, do not manage them
net = openstack::Network(provider=provider, project=project, name=network_name, managed=false)
subnet = openstack::Subnet(provider=provider, project=project, network=net, dhcp=true, managed=false,
                           name=subnet_name, network_address=network_address)

# Define the virtual machine
testhost = openstack::Host(provider=provider, project=project, key_pair=ssh_key, name="testhost",
                     image=image_id, os=os, flavor=flavor_name, user_data="", subnet=subnet)
testhost2 = openstack::Host(provider=provider, project=project, key_pair=ssh_key, name="testhost2",
                     image=image_id, os=os, flavor=flavor_name, user_data="", subnet=subnet)
```
project.yml:
```
name: snippet test
modulepath: [libs,/tmp/tmpmqtlmkgr]
downloadpath: /tmp/tmpmqtlmkgr
version: 1.0
repo: ['https://github.com/inmanta/']
```

I just took the example in `openstack` and added the security group creation

**Comment by bartv** [comment]
Found it.

You define the SG for each instance. Because of the index, the compiler will only create one instance for the SG but it will add the rules twice. The handler cannot handle this. The first part of the fix is the following:

```
diff --git a/plugins/__init__.py b/plugins/__init__.py
index 1634b32..f54a006 100644
--- a/plugins/__init__.py
+++ b/plugins/__init__.py
@@ -293,6 +294,7 @@ class SecurityGroup(OpenstackResource):
     @staticmethod
     def get_rules(exporter, group):
         rules = []
+        dedup = set()
         for rule in group.rules:
             json_rule = {"protocol": rule.ip_protocol,
                          "direction": rule.direction}
@@ -321,7 +323,12 @@ class SecurityGroup(OpenstackResource):
             except Exception:
                 pass
 
-            rules.append(json_rule)
+            key = tuple(sorted(json_rule.items()))
+            if key not in dedup:
+                dedup.add(key)
+                rules.append(json_rule)
+            else:
+                LOGGER.warning("A duplicate rule exists in security group %s", group.name)
 
         return rules
```

I am also thinking about adding a validation function as plugin that will raise a compiler error when it detects this.

**Comment by bartv** [comment]
Fixed in f74b6da14f2eff7f36f58b49cad143c5454b24e6 / 0.2.3.4
